zerver/tests/test_user_topics.py

```diff
@@ -641,3 +643,189 @@ def test_unmuted_topic_remove_invalid(self) -> None:
+        )
+        self.assertEqual(user_ids, {hamlet.id})
+
+    def test_automatically_follow_topic_on_participation(self) -> None:
+        hamlet = self.example_user("hamlet")
+        cordelia = self.example_user("cordelia")
+        iago = self.example_user("iago")
+        othello = self.example_user("othello")
+        aaron = self.example_user("aaron")
+        realm = hamlet.realm
+        stream = get_stream("Verona", realm)
+        topic_name = "teST topic"
+
+        stream_topic_target = StreamTopicTarget(
+            stream_id=stream.id,
+            topic_name=topic_name,
+        )
+
+        user_ids = stream_topic_target.user_ids_with_visibility_policy(
+            UserTopic.VisibilityPolicy.FOLLOWED
+        )
+        self.assertEqual(user_ids, set())
+
+        def set_automatically_follow_topic_on_participation(user: UserProfile) -> None:
+            do_change_user_setting(
+                user,
+                "automatically_follow_topics_policy",
+                UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_PARTICIPATION,
+                acting_user=None,
+            )
+
+        # - For hamlet, cordelia, iago and othello:
+        #   'automatically_follow_topics_policy' set to 'AUTOMATICALLY_FOLLOW_TOPICS_ON_PARTICIPATION'
+        # - For aaron:
+        #   'automatically_follow_topics_policy' NOT set to 'AUTOMATICALLY_FOLLOW_TOPICS_ON_PARTICIPATION'.
+        for user in [hamlet, cordelia, iago, othello]:
+            set_automatically_follow_topic_on_participation(user)
+        do_change_user_setting(
+            aaron,
+            "automatically_follow_topics_policy",
+            UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_I_START,
+            acting_user=None,
+        )
+
+        # Participation includes:
+        # * Sending a message
+        # * Reacting to a message
+        # * Participating in a poll
+        # * Editing a todo list
+        # Hamlet sends a message. DO automatically follow the topic.
+        # Aaron sends a message. DON'T automatically follow the topic.
+        message_id = self.send_stream_message(
+            hamlet, stream_name=stream.name, topic_name=topic_name
+        )
+        self.send_stream_message(aaron, stream_name=stream.name, topic_name=topic_name)
+        user_ids = stream_topic_target.user_ids_with_visibility_policy(
+            UserTopic.VisibilityPolicy.FOLLOWED
+        )
+        self.assertEqual(user_ids, {hamlet.id})
+
+        # Cordelia reacts to a message. DO automatically follow the topic.
+        # Aaron reacts to a message. DON'T automatically follow the topic.
+        def add_reaction(user: UserProfile) -> None:
+            reaction_info = {
+                "emoji_name": "smile",
+            }
+            endpoint = f"/api/v1/messages/{message_id}/reactions"
+            result = self.api_post(user, endpoint, reaction_info)
+            self.assert_json_success(result)
+
+        add_reaction(cordelia)
+        add_reaction(aaron)
```

Comment:
Might be more convenient to use `/actions/reactions.py/check_add_reaction` than calling the endpoint here.

Reply:
Used `check_add_reaction`